DESIGN.md

```diff
@@ -0,0 +1,471 @@
+                             +-------------v--------------+
+                             |     Manager Library        |
+                             |     Job Controller         |
+                             +------+------+--------------+
+                                           | 
+                             +-------------v-------------+
+                             |    Job Library (low level)|
+                             |    Core Job functionality |
+                             +------+------------+-------+
+                                    |            |
+                           +--------v-----+    +-v-----------+
+                           | Process Ctl  |    | Cgroup Ctl  |
+                           |(Start/Stop/  |    | Limit       |
+                           |Status/Stream)|    |CPU/Memory/IO|
+                           +--------------+    +-------------+
+```
+---
+
+## Scope
+
+### Included
+
+- Start/stop/query status of arbitrary processes. Note: Will **always** stream messages in chunks from the beginning of the output
+- Stream process stdout/stderr. Data will persist across server reboots.
+- Semi-persistent storage of job stream info
+- Terminate child processes correctly, no orphans or zombies (done via `cgroups` and `setpgid`)
+- Set resource limits (CPU, memory, I/O) using cgroups v2 (manual implementation)
+- Secure gRPC communication via mTLS (strong cipher suites)
+- Simple authorization via client CA
+- CLI utility to interact with server
+- Minimal Unit testing
+- Happy path and common error scenarios integration testing
+- Basic end to end tests
+- Command denylist for normal users, but will run as `nobody` user
+- Same command denylist for admin users and will run as `1000:1000` user
+- Job isolation via cgroups and user/group based access control
+- Job metadata management (in memory only for now)
+- Job ID generation (unique, lexicographically sortable)
+- Logging 
+
+### Excluded
+  (Due to time constraints)
+- Horizontal scalability. I will only implement this in one instance of compute.
+- Availability (multiple instances of compute) - This is a single source of failure. If the server goes down, the service goes down or if the server is overloaded, cli experience will suffer.
+- Containerization of compute (ie Kubernetes). Using kubernetes is a clean way to implement distributed compute. 
+- Persistent storage of client CAs in a database like sqlite
+- Configuration - Will use hardcoding as much as possible for very simple configurations server/client  port, limits, denylist and anything else that could come up.
+- DDOS Prevention. No rate limiting middleware will be implemented
+
+---
+
+## Security Considerations
+
+- **mTLS** for mutual authentication and transport encryption
+  - TLS 1.3 only
+  - Strong cipher suite (TLS\_AES\_256\_GCM\_SHA384 preferred)
+- **Authorization policy**
+  - In the client certificate itself, `CN` will be the `username` and `O` will be the `role-group`
```

Comment:
After reading through this a couple times I'm still not totally clear on the authorization model. Maybe having both roles _and_ user groups is more complicated than what we need here? I'd welcome any opportunity to simplify.